lib/Doctrine/ODM/MongoDB/Iterator/CachingIterator.php

```diff
@@ -94,9 +98,14 @@ public function key()
     /** @see http://php.net/iterator.next */
     public function next(): void
     {
-        if (! $this->iteratorExhausted) {
-            $this->getIterator()->next();
+        if ($this->iterator !== null) {
+            $this->iterator->next();
             $this->storeCurrentItem();
+            $this->iteratorAdvanced = true;
+
+            if (! $this->iterator->valid()) {
+                $this->iterator = null;
+            }
```

Comment:
Related to the `storeCurrentItem` refactoring:

```suggestion
        $this->iterator?->next();
        $this->storeCurrentItem();
        $this->iteratorAdvanced = true;
```

Reply:
I've applied this change partially, as I think having an `if` block for the case of non-exhausted iterator is more explicit for the phase of reading the iterator.